crates/rust-analyzer/tests/heavy_tests/main.rs

```diff
@@ -715,3 +715,183 @@ pub fn foo(_input: TokenStream) -> TokenStream {
+                server.doc_id("src/lib.rs"),
+                Position::new(2, 12),
+            ),
+            work_done_progress_params: Default::default(),
+        },
+        json!({
+            "actions": [{
+                "commands": [{
+                    "arguments": [
+                      "file:///[..]src/lib.rs",
+                      { "character": 7, "line": 2 },
+                      []
+                    ],
+                    "command": "rust-analyzer.showReferences",
+                    "title": "0 implementations",
+                    "tooltip": "Go to implementations"
+                }]
+            }],
+            "contents": { "kind": "markdown", "value": "```rust\nfoo\n```\n\n```rust\nstruct NoImpl\n```" },
+            "range": { "end": { "character": 13, "line": 2 }, "start": { "character": 7, "line": 2 } }
+        })
+    );
+}
+
+#[test]
+fn test_client_does_not_support_hover_actions() {
+    if skip_slow_tests() {
+        return;
+    }
+
+    let server = Project::with_fixture(
+        r#"
+//- Cargo.toml
+[package]
+name = "foo"
+version = "0.0.0"
+
+//- src/lib.rs
+struct Foo(u32);
+
+struct NoImpl(u32);
+
+impl Foo {
+    fn new() -> Self {
+        Self(1)
+    }
+}
+"#,
+    )
+    .with_config(|config| {
+        config.client_caps.hover_actions = false;
+    })
+    .server();
+
+    server.wait_until_workspace_is_loaded();
+
+    // has 1 implementation
+    server.request::<HoverRequest>(
+        HoverParams {
+            text_document_position_params: TextDocumentPositionParams::new(
+                server.doc_id("src/lib.rs"),
+                Position::new(0, 9),
+            ),
+            work_done_progress_params: Default::default(),
+        },
+        json!({
+            "contents": { "kind": "markdown", "value": "```rust\nfoo\n```\n\n```rust\nstruct Foo\n```" },
+            "range": { "end": { "character": 10, "line": 0 }, "start": { "character": 7, "line": 0 } }
+        })
+    );
+
+    // no hover
+    server.request::<HoverRequest>(
+        HoverParams {
+            text_document_position_params: TextDocumentPositionParams::new(
+                server.doc_id("src/lib.rs"),
+                Position::new(1, 0),
+            ),
+            work_done_progress_params: Default::default(),
+        },
+        json!(null),
+    );
+
+    // no implementations
+    server.request::<HoverRequest>(
+        HoverParams {
+            text_document_position_params: TextDocumentPositionParams::new(
+                server.doc_id("src/lib.rs"),
+                Position::new(2, 12),
+            ),
+            work_done_progress_params: Default::default(),
+        },
+        json!({
+            "contents": { "kind": "markdown", "value": "```rust\nfoo\n```\n\n```rust\nstruct NoImpl\n```" },
+            "range": { "end": { "character": 13, "line": 2 }, "start": { "character": 7, "line": 2 } }
+        })
+    );
+}
```

Comment:
I'd rather not add heavy tests for this functionality:

* the cost of a heavy test is high, as it spawns an external process
* the additional confidence it provides is low, as it only verifies the JSON we produce, and not how the other side interprets the JSON. 

So, almost all code in `handlers.rs` and `conv.rs` is deliberately not tested, and we just rely on types and manual testing for correctness. 

So far, it worked ok -- the regressions I remember usually involve bugs in https://github.com/Microsoft/vscode-languageserver-node/

Reply:
I'm not against the deleting. Just one consideration:
> additional confidence it provides is low, as it only verifies the JSON we produce,

I do not think so. Such tests let us to ensure that we produce **valid** JSON.

For example, these tests helped me to find and fix a problem: with disabled hoverActions client capability, the server was still adding empty actions field to the Hover response. Of course, LSP defines that an unknown property should be ignored, but it is not always the case.

Reply:
I still think it makes sense to delete. I agree that this can help to find some bugs, probably quite a few of them. But, over the lifetime of the project, the cost of running&maintaining these test seems like it would be higher than the additional benefit.

Reply:
(as a more general design note, I find that for ironing bugs on the integration boundary the most efficient strategy is release trains (so that folks can try things out in various real world configurations) and speedy release process (so that you can submit hotfix and make a new release quckly))

Reply:
yea, this strategy should work well for such an active project :)